### PR TITLE
202409 load rules

### PR DESCRIPTION
**Pull request by petermarshallio**
A new notebook focused on using load rules.

To facilitate this, a new docker profile has been created that spins up a second historical, and which only spins up when `druid-jupyter-tiered-hist` profile is used.

The notebook walks through using interval, period, and forever load rules, shows how rule order affects processing, and shows how to leave some data in deep storage and unloaded.

### CODE REVIEW

**Review by hevansDev — CHANGES_REQUESTED** [review]

File: notebooks/03-query/20-tiering-historicals.ipynb [hunk 1]
```diff
@@ -0,0 +1,784 @@
...
+    "  ~\n",
+    "  ~ Unless required by applicable law or agreed to in writing,\n",
+    "  ~ software distributed under the License is distributed on an\n",
+    "  ~ \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY\n",
+    "  ~ KIND, either express or implied.  See the License for the\n",
+    "  ~ specific language governing permissions and limitations\n",
+    "  ~ under the License.\n",
+    "  -->\n",
+    "\n",
+    "[Service tiering](https://druid.apache.org/docs/latest/operations/mixed-workloads#service-tiering) provides administrators the ability to provide cluster resources suited to different performance and storage requirements, such as isolating heavy queries involving complex subqueries or large result from high priority, interactive queries.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/20-tiering-historicals.ipynb:28]:
```suggestion
    "As an Apache Druid cluster administrator, you can use [Service tiering](https://druid.apache.org/docs/latest/operations/mixed-workloads#service-tiering) to provision cluster resources according to various performance and storage requirements. For example, you can isolate resource-intensive queries with complex subqueries and large result sets from high priority, interactive queries.\n",
```

**Review by hevansDev — APPROVED** [review]

File: notebooks/03-query/20-tiering-historicals.ipynb [hunk 2]
```diff
@@ -0,0 +1,790 @@
...
+   "metadata": {},
+   "source": [
+    "Review the following retention rule.\n",
+    "\n",
+    "- An [interval load rule](https://druid.apache.org/docs/latest/operations/rule-configuration#interval-load-rule) (`loadByInterval`) covers 10 years of data before 1st January 2016, and requests one cached replica on both the `_default_tier` and `slow` tiers.\n",
+    "- `loadForever` requests that all data be cached on the `slow` tier.\n",
+    "\n",
+    "Each segment is checked against the rules in order when the decision is made as to where it must be cached.\n",
+    "\n",
+    "Run the cell to commit it to the database."
```
Comment (techdocsmith) [inline notebooks/03-query/20-tiering-historicals.ipynb:436]: This needs revision. The code accomplishes the correct task, but based upon this description I was confounded.

Discrepencies:

Line 431 directly opposes line 490, which correctly describes the result:
> Data for 2015 (`loadByInterval`) is only cached on the `slow` tier.

Line 432 directly opposes Line 491, which correctly describes the result:
> All other data (`loadForever`) is only available on the `_default_tier` tier.

* Need to say that Druid applies the load rules in order first
* Need to say that the interval load rule assigns the interval of 10 years up to 1 Jan 2016 to the _slow_ tier
* Need to say that somehow the remaining data gets assigned to the default tier based upon the loadForever rule. Like if data is already handled by the interval rule, it won't be assigned here.

File: notebooks/03-query/20-tiering-historicals.ipynb [hunk 3]
```diff
@@ -0,0 +1,790 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "37e887eb-f2b9-4014-b2bd-e2e0186a56f3",
+   "metadata": {},
+   "source": [
+    "### Load data according to age\n",
+    "\n",
+    "Run the cell below to ingest some data for the current year.\n",
+    "\n",
+    "The TIME_EXTRACT function is used to fake data for this year by calculating the shift between the source timestamp and today's date."
```
Comment (techdocsmith) [inline notebooks/03-query/20-tiering-historicals.ipynb:535]:
```suggestion
    "The TIME_EXTRACT function calculates the shift between the source timestamp and today's date to fake data for this year."
```

File: notebooks/03-query/20-tiering-historicals.ipynb [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f976d910-d4e9-43d6-ad8c-4f03bc4dd762",
+   "metadata": {},
+   "source": [
+    "Review the following retention load rules configuration, then run the cell to apply it to the table.\n",
+    "\n",
+    "- `loadByPeriod`, giving a [period load rule](https://druid.apache.org/docs/latest/operations/rule-configuration/#period-load-rule) covering data newer than one year old, requesting one cached replica on both the `_default_tier` and `slow` tiers.\n",
+    "- `loadForever`, requesting all data to be cached on the `slow` tier."
```
Comment (techdocsmith) [inline notebooks/03-query/20-tiering-historicals.ipynb:580]:
```suggestion
    "The following retention load rules configuration creates load rules as follows:\n",
    "\n",
    "- `loadByPeriod`, creates a [period load rule](https://druid.apache.org/docs/latest/operations/rule-configuration/#period-load-rule) for one cached replica of data newer than one year on both the `_default_tier` and `slow` tiers.\n",
    "- `loadForever`, creates a rule to cache all remaining data on the `slow` tier."
```
i think we should be careful about "load forever" requesting "all data" on the slow tier. It's really only applying to the data that was not handled by the first load rule, right?

File: notebooks/03-query/20-tiering-historicals.ipynb [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+   "id": "1f8ac22c-9fd5-4209-86c5-41fd7b4431cb",
+   "metadata": {},
+   "source": [
+    "Notice that, depending on the period of time they cover, some table segments are loaded onto historicals and some are not.\n",
+    "\n",
+    "The order of the rules means:\n",
+    "\n",
+    "1. Data younger than a year is loaded to `_default_tier` historicals.\n",
+    "2. Data covering 2015 is cached on `slow`-tier historicals.\n",
+    "3. No other data is loaded."
```
Comment (techdocsmith) [inline notebooks/03-query/20-tiering-historicals.ipynb:718]:
```suggestion
    "1. Druid loads data younger than a year onto `_default_tier` historicals.\n",
    "2. Druid loads data covering 2015 onto `slow`-tier historicals.\n",
    "3. No other data is loaded."
```

**Review by techdocsmith — CHANGES_REQUESTED** [review]
This is a great guide. There are some problems with the description for first foray into dividing data amongst tiers, but the code is OK.

File: notebooks/03-query/20-tiering-historicals.ipynb (hunk 2 above)
Comment (petermarshallio) [reply notebooks/03-query/20-tiering-historicals.ipynb]: THANK YOU! I didn't spot that! Updated.

**Pull request merged**